src/coprocessor/dag/expr/builtin_time.rs

```diff
@@ -418,6 +415,42 @@ impl ScalarFunc {
     ) -> Result<Option<Cow<'a, MyDuration>>> {
         Ok(Some(Cow::Owned(MyDuration::zero())))
     }
+
+    #[inline]
+    pub fn add_date_and_duration<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let d0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let d1: Cow<'a, MyDuration> = try_opt!(self.children[1].eval_duration(ctx, row));
+        let add = match d0.to_nanos().checked_add(d1.to_nanos()) {
+            Some(result) => result,
+            None => return Err(box_err!("add duration {} and duration {} overflow", d0, d1)),
+        };
+        let res = MyDuration::from_nanos(add, d0.fsp().max(d1.fsp()) as i8)?;
+        Ok(Some(Cow::Owned(res.to_string().into_bytes())))
+    }
+
+    #[inline]
+    pub fn add_date_and_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let arg0: Cow<'a, MyDuration> = try_opt!(self.children[0].eval_duration(ctx, row));
+        let arg1: Cow<'a, [u8]> = try_opt!(self.children[1].eval_string(ctx, row));
+        let s: &str = box_try!(::std::str::from_utf8(arg1.as_ref()));
+        let arg1 = MyDuration::parse(s.as_bytes(), Time::parse_fsp(s))?;
```

Comment:
could we use `MyDuration::parse(arg1, Time::parse_fsp(s))?` directly?

Reply:
The signature of parse method is `pub fn parse(mut s: &[u8], fsp: i8) -> Result<Duration>`. 
I think the following three solutions are feasible：
1. ` MyDuration::parse(&arg1, Time::parse_fsp(s))?;`
2. ` MyDuration::parse(arg1.as_ref(), Time::parse_fsp(s))?;`
3. ` MyDuration::parse(s.as_bytes(), Time::parse_fsp(s))?;`

Maybe the solution ONE is better?

Reply:
Ok @koushiro